File: trees/94_binary_tree_inorder_traversal/94_binary_tree_inorder_traversal.py
"""
LeetCode 94. Binary Tree Inorder Traversal
Link: https://leetcode.com/problems/binary-tree-inorder-traversal/
Difficulty: Easy
Primary: Tree
Tags: [tree, binary-tree, dfs, recursion, stack]
Solved: 2025-11-13
"""

from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:

    # --- Approach 1: Depth First Search ---
    # Time: O(n), Space: O(n)
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []

        def dfs(node):
            if not node:
                return
            
            dfs(node.left)
            res.append(node.val)
            dfs(node.right)
        
        dfs(root)
        return res
    
    # --- Approach 2: Iterative DFS ---
    # Time: O(n), Space: O(n)
    def inorderTraversal_iterative(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        stack = []
        cur = root

        while cur or stack:
            while cur:
                stack.append(cur)
                cur = cur.left
            
            cur = stack.pop()
            res.append(cur.val)
            cur = cur.right
        
        return res    
    

# --- Example Test ---
if __name__ == "__main__":
    solution = Solution()

    root1 = TreeNode(1, None, TreeNode(2, TreeNode(3), None))
    print(solution.inorderTraversal(root1))  # Output: [1, 3, 2]

File: trees/94_binary_tree_inorder_traversal/94_binary_tree_inorder_traversal.java
/*
LeetCode 94. Binary Tree Inorder Traversal
Link: https://leetcode.com/problems/binary-tree-inorder-traversal/
Difficulty: Easy
Primary: Tree
Tags: [tree, binary-tree, dfs, recursion, stack]
Solved: 2025-11-13
*/

import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {

    // --- Approach 1: Depth First Search ---
    // Time: O(n), Space: O(n)
    private List<Integer> res;

    public void dfs(TreeNode node) {
        if (node == null) {
            return;
        }

        dfs(node.left);
        res.add(node.val);
        dfs(node.right);
    }

    public List<Integer> inorderTraversal_dfs(TreeNode root) {
        res = new ArrayList<>();
        dfs(root);
        return res;
    }


    // --- Approach 2: Iterative DFS ---
    // Time: O(n), Space: O(n)
    public List<Integer> inorderTraversal_iterativeDfs(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        TreeNode cur = root;
        while (cur != null || !stack.isEmpty()) {
            while (cur != null) {
                stack.push(cur);
                cur = cur.left;
            }
            cur = stack.pop();
            res.add(cur.val);
            cur = cur.right;
        }

        return res;
    }


    // --- Example Test ---
    public static void main(String[] args) {
        Solution sol = new Solution();
        TreeNode root1 = new TreeNode(1, null, new TreeNode(2, new TreeNode(3), null));
        System.out.println(sol.inorderTraversal_dfs(root1));  // Output: [1, 3, 2]
    }
    
}